SimPEG/regularization/rotated.py

```diff
@@ -0,0 +1,241 @@
+                    f"of active cells {n_cells}, or the number of mesh cells {mesh.n_cells}. "
+                )
+        if np.any(alphas < 0):
+            raise ValueError("`alpha` must be non-negative")
+        anis_alpha = alphas
+
+        if reg_dirs is not None:
+            reg_dirs = validate_ndarray_with_shape(
+                "reg_dirs",
+                reg_dirs,
+                shape=[(mesh.dim, mesh.dim), ("*", mesh.dim, mesh.dim)],
+                dtype=float,
+            )
+            if reg_dirs.shape == (mesh.dim, mesh.dim):
+                reg_dirs = np.tile(reg_dirs, (mesh.n_cells, 1, 1))
+            if reg_dirs.shape[0] != mesh.n_cells:
+                # check if I need to expand from active cells to all cells (needed for discretize)
+                if (
+                    reg_dirs.shape[0] == n_active_cells
+                    and self.active_cells is not None
+                ):
+                    reg_dirs_temp = np.zeros((mesh.n_cells, mesh.dim, mesh.dim))
+                    reg_dirs_temp[self.active_cells] = reg_dirs
+                    reg_dirs = reg_dirs_temp
+                else:
+                    raise IndexError(
+                        f"`reg_dirs` first dimension, {reg_dirs.shape[0]}, must be either number "
+                        f"of active cells {n_cells}, or the number of mesh cells {mesh.n_cells}. "
+                    )
+            # check orthogonality?
+            if ortho_check:
+                eye = np.eye(mesh.dim)
+                for i, M in enumerate(reg_dirs):
+                    if not np.allclose(eye, M @ M.T):
+                        raise ValueError(f"Matrix {i} is not orthonormal")
+            # create a stack of matrices of dir @ alphas @ dir.T
+            anis_alpha = np.einsum("ink,ik,imk->inm", reg_dirs, anis_alpha, reg_dirs)
+            # Then select the upper diagonal components for input to discretize
+            if mesh.dim == 2:
+                anis_alpha = np.stack(
+                    (
+                        anis_alpha[..., 0, 0],
+                        anis_alpha[..., 1, 1],
+                        anis_alpha[..., 0, 1],
+                    ),
+                    axis=-1,
+                )
+            elif mesh.dim == 3:
+                anis_alpha = np.stack(
+                    (
+                        anis_alpha[..., 0, 0],
+                        anis_alpha[..., 1, 1],
+                        anis_alpha[..., 2, 2],
+                        anis_alpha[..., 0, 1],
+                        anis_alpha[..., 0, 2],
+                        anis_alpha[..., 1, 2],
+                    ),
+                    axis=-1,
+                )
+        self._anis_alpha = anis_alpha
+
+    # overwrite the call, deriv, and deriv2...
+    def __call__(self, m):
+        G = self.cell_gradient
+        M_f = self.W
+        r = G @ (self.mapping * (self._delta_m(m)))
+        return 0.5 * r @ M_f @ r
```

Comment:
Here we only have one `M_f` with `M_f = self.W`, don't we miss another to have W^TW ?

Reply:
(same comment for deriv and deriv2)

Reply:
Actually no, the way it works here is that `M_f=W.T@W` in the standard way of describing the regularization. Thus since I use discretize to construct the anisotropic inner product matrix. So I just set `W=M_f` since there’s a lot of machinery already hooked into updated the weighting matrix on things like weighting function changes.

Reply:
Thanks. I do admit I find this extremely confusing and important to document here.

Reply:
`M_f = self.W`; then should be `M_f = self.WTW` for clarity